### ISSUE TITLE
ltnews31, ptex-p3.8.3

### ISSUE THREAD

**Issue by aminophen** [clerkma/ptex-ng#47]
## pdfTeX primitives

As stated in [ltnews31](https://www.latex-project.org/news/latex2e-news/ltnews31.pdf), LaTeX will require many pdfTeX-derived primitives.

I see the following issues:

* `\expanded`: current ptex-ng returns `\show\expanded` => `> \expanded=\jobname.`
* `\ifincsname`: undefined (added in e-pTeX 190709)
* `\ifpdfprimitive`, `\pdfprimitive`: undefined (added in e-pTeX 150805; many bugs are fixed in 2017; search "pdfprimitive" through web2c/eptexdir/ChangeLog)
* `\pdfshellescape`: undefined (added in e-pTeX 141108)

## pTeX primitives

Please see changes in pTeX r53726-r53727; current ptex-ng returns wrong values in fetching `\kansujichar`.  The below is a test suite. (expected output is shown in comments)

``` tex
%% block 1
\showthe\kansujichar1\relax % => 19968.
\showthe\dimen200\relax     % => 0.0pt.
\showthe\kansujichar1\relax % => 19968.
\showthe\skip200\relax      % => 0.0pt.
\showthe\kansujichar1\relax % => 19968.
\showthe\count200\relax     % => 0.
\showthe\kansujichar1\relax % => 19968.

%% block 2
\showthe\kansujichar10\relax % => ! Invalid KANSUJI number (10). % => 10.
\showthe\dimen200\relax      % => 0.0pt.
\showthe\kansujichar10\relax % => ! Invalid KANSUJI number (10). % => 10.
\showthe\skip200\relax       % => 0.0pt.
\showthe\kansujichar10\relax % => ! Invalid KANSUJI number (10). % => 10.
\showthe\count200\relax      % => 0.
\showthe\kansujichar10\relax % => ! Invalid KANSUJI number (10). % => 10.

\end
```

**Comment by clerkma** [comment]
Thanks.  Current status is listed here: 
- `\show\expanded`: fixed
- `\kansujichar10`: fixed
- `\pdfshellescape`: I added an alias (`\shellescape` to `\pdfshellescape`)
- `\ifincsname`: TODO
- `\ifpdfprimitive`: TODO
- `\pdfprimitive`: TODO

For the last three primitives, I still need one day or two.

**Comment by clerkma** [comment]
finished with [9d276ee5cc5bfec5f2a050c4c1a53ea7e763b8b8](https://github.com/clerkma/ptex-ng/commit/9d276ee5cc5bfec5f2a050c4c1a53ea7e763b8b8).

**Comment by aminophen** [comment]
Thanks, all of them are working as expected!

I noticed some additional differences between euptex and ptex-ng:

* \fontcharwd etc. for char<256: See https://github.com/TeX-Live/texlive-source/commit/2f118d2c4c0d731ef041a26cea7fc80d48e35365. MWE:

``` tex
%#!euptex
\jfont\x=upjisr-h \x
\showthe\fontcharwd\jfont`°\relax % => should be "10.0pt"
\showthe\fontcharwd\jfont"B7\relax % => should be "5.0pt"
\showthe\fontcharwd\jfont`ー\relax % => 10.0pt
\showthe\fontcharwd\jfont`、\relax % => 5.0pt
\showthe\fontcharwd\jfont`（\relax % => 5.0pt
\showthe\fontcharwd\tfont`漢\relax % => 10.0pt
\bye
```

* New specification with \fontcharwd etc. and \iffontchar, introduced in e-pTeX 190709: See https://github.com/TeX-Live/texlive-source/commit/8a39f4e259e5adee484708aa18562ea2d7b5a3d4. Excerpt from ChangeLog:

```
	* eptex.ech: Revise \iffontchar and \fontchar?? to treat
	negative value as char_type.
	  - \iffontchar (Japanese font f) (number c):
	    - c >= 0: true iff is_char_kanji(c)
	    - c < 0: true iff f has char_type -(c+1)
	  - \fontcharwd (Japanese font f) (number c):
	    - c >= 0: return the width of character c
	    - c < 0: return the width of char_type -(c+1)
```

**Comment by clerkma** [comment]
fixed in [ddd1535546585af42036e73db7e5e7896aaa544e](https://github.com/clerkma/ptex-ng/commit/ddd1535546585af42036e73db7e5e7896aaa544e).

**Comment by aminophen** [comment]
Sorry for late reply, confirmed thanks!
